### PR TITLE
More GPI cleanup

### PR DESCRIPTION
**Pull request by ktbarrett**
Some more small cleanup commits. Will be working on adding interfaces to the GPI to register startup, shutdown, and cleanup callbacks. This will be used to replace the embed API.

### CODE REVIEW

**Comment by ktbarrett** [comment]
Works just fine on my machine, so I guess I keep playing around with it.

File: cocotb/share/lib/embed/gpi_embed.cpp [hunk 1]
```diff
@@ -96,9 +96,17 @@ static void set_program_name_in_venv(void) {
         return;
     }
 
-    wcsncpy(venv_path_w, Py_DecodeLocale(venv_path, NULL),
-            sizeof(venv_path_w) / sizeof(wchar_t));
+    auto venv_path_w_temp = Py_DecodeLocale(venv_path, NULL);
+    if (venv_path_w_temp == NULL) {
+        LOG_ERROR(
+            "Unable to set Python Program Name using virtual environment. "
+            "Decoding error in virtual environment path.");
```
Comment (cmarqu) [inline cocotb/share/lib/embed/gpi_embed.cpp:103]: Print the path here?

File: cocotb/share/lib/gpi/GpiCommon.cpp [hunk 2]
```diff
@@ -124,7 +124,7 @@ void gpi_embed_init(int argc, char const *const *argv) {
 }
 
 void gpi_embed_end() {
-    embed_sim_event(SIM_FAIL, "Simulator shutdown prematurely");
+    embed_sim_event("Simulator shutdown prematurely");
```
Comment (cmarqu) [inline cocotb/share/lib/gpi/GpiCommon.cpp:127]:
```suggestion
    embed_sim_event("Simulator shut down prematurely");
```

File: cocotb/share/lib/embed/gpi_embed.cpp [hunk 3]
```diff
@@ -261,30 +256,22 @@ extern "C" COCOTB_EXPORT int _embed_sim_init(int argc,
...
+        }
+    })
     for (int i = 0; i < argc; i++) {
-        // Decode, embedding non-decodable bytes using PEP-383. This can only
-        // fail with MemoryError or similar.
-        auto argv_item = PyUnicode_DecodeLocale(argv[i], "surrogateescape");
-        if (!argv_item) {
-            // LCOV_EXCL_START
-            PyErr_Print();
+        auto arg = Py_DecodeLocale(argv[i], NULL);
```
Comment (marlonjames) [inline cocotb/share/lib/embed/gpi_embed.cpp:266]: Does changing from `PyUnicode_DecodeLocale(argv[i], "surrogateescape")` to `Py_DecodeLocale(argv[i], NULL)` change cocotb behavior at all?
@eric-wieser 

From the docs for `PyUnicode_DecodeLocale()`:
> Changed in version 3.7: The function now also uses the current locale encoding for the `surrogateescape` error handler, except on Android. Previously, [Py_DecodeLocale()](https://docs.python.org/3/c-api/sys.html#c.Py_DecodeLocale) was used for the `surrogateescape`, and the current locale encoding was used for `strict`.

Comment (ktbarrett) [reply cocotb/share/lib/embed/gpi_embed.cpp:266]: I'm less convinced this commit is beneficial, looking over it now. The only real value in calling `PySys_SetArgv` is to set argv *before* initialization to assist in determining the Python executable and environment. I'll rebase it out tonight unless there are objections.

File: cocotb/__init__.py [hunk 4]
```diff
@@ -322,26 +322,15 @@ def _initialise_testbench_(argv_):
...
-    # SIM_INFO = 0
-    SIM_TEST_FAIL = 1
-    SIM_FAIL = 2
     from cocotb.result import SimFailure
 
-    if level is SIM_TEST_FAIL:
-        scheduler.log.error("Failing test at simulator request")
-        scheduler._finish_test(
-            AssertionError(f"Failure from external source: {message}")
-        )
```
[the remark this answers is not in the document]
Comment (ktbarrett) [reply cocotb/__init__.py:336]: We still have a "deprecated" public alias.

**Comment by marlonjames** [comment]
Triggering tests again

File: documentation/source/building.rst [hunk 5]
```diff
@@ -242,7 +242,7 @@ PyGPI
     The entry module must have the following additional functions defined.
     These additional requirements on the entry module may be relaxed over time.
 
-    * ``_sim_event(level: int) -> None``
+    * ``_sim_event(message: str) -> None``
```
Comment (marlonjames) [inline documentation/source/building.rst:245]: Might consider adding a `.. versionchanged::1.8` note here, to start noting changes over time.

File: documentation/source/building.rst [hunk 6]
```diff
@@ -242,7 +242,7 @@ PyGPI
     The entry module must have the following additional functions defined.
     These additional requirements on the entry module may be relaxed over time.
```
Comment (ktbarrett) [inline documentation/source/building.rst:243]: .. versionchanged:: 1.8
    ``level`` argument no longer necessary.

File: documentation/source/building.rst [hunk 7]
```diff
@@ -240,9 +240,10 @@ PyGPI
     * ``entry_function(argv: List[str]) -> None``
 
     The entry module must have the following additional functions defined.
-    These additional requirements on the entry module may be relaxed over time.
+.. versionchanged:: 1.8
+    ``level`` argument is no longer passed, it is assumed to be `SIM_FAIL` (2).
```
Comment (marlonjames) [inline documentation/source/building.rst:244]:
```suggestion
    .. versionchanged:: 1.8
        ``level`` argument is no longer passed to ``_sim_event``, it is assumed to be `SIM_FAIL` (2).
```

**Review by marlonjames — APPROVED** [review]
Indent fix for docs.

**Pull request merged**